File: tools/inkvec_sr/clean.py
"""The pre-pass itself: upscale, halve, and put the colours back.

Measured during development. The short version is that the
upscaler removes about two thirds of a JPEG's error energy -- not because it was
trained to, it was not, but because it was fitted on flat-colour logo art and
ringing is not on that manifold, so it cannot represent the damage.

Three steps, and each is here for a measured reason.

**Upscale x4.** This is what actually cleans. Artefact gain -- what survives the
network relative to what went in -- is 0.36 across seven families.

**Box-downsample x2.** Not a denoiser, whatever it looks like. A 2x2 box has its
only transfer zero at Nyquist, and the DCT grid puts JPEG damage at f = 1/8 and
below, which it passes at 0.974 of amplitude; half the error survives and the
surviving half is the *more* coherent half. It is here because x2 is the useful
output scale, not because it cleans.

**Recolour.** The network costs about 0.6 dE00 on clean input too, so that cost
is not about JPEG. Split by region against bicubic as control, it is 0.60 on flat
interiors where bicubic scores 0.03, and 1.57 on edges where bicubic scores 8.63:
each resampler is right about the half the other is wrong about. So take geometry
from the network and colour from the source. What survives is the edge column,
which is a sub-pixel boundary disagreement and the tracer's problem, not this
module's.
"""
from __future__ import annotations

import numpy as np

#: Interior residual above which an input is treated as degraded. Measured at
#: 0.000 on an exact-coverage clean intake and 1.509 on JPEG q50, so anywhere in
#: between is safe; this sits low enough to catch milder damage.
DEGRADED_RESIDUAL = 0.5


def box_downsample(a: np.ndarray, factor: int = 2) -> np.ndarray:
    """Exact box average by an integer factor."""
    f = factor
    h, w = a.shape[0] // f * f, a.shape[1] // f * f
    b = a[:h, :w]
    return b.reshape(h // f, f, w // f, f, -1).mean(axis=(1, 3))


def _flat_mask(img: np.ndarray, tol: float) -> np.ndarray:
    """True where the 3x3 neighbourhood spans less than `tol`."""
    g = img.mean(axis=2)
    lo = hi = g
    for dy in (-1, 0, 1):
        for dx in (-1, 0, 1):
            r = np.roll(np.roll(g, dy, 0), dx, 1)
            lo, hi = np.minimum(lo, r), np.maximum(hi, r)
    flat = (hi - lo) < tol
    flat[:1, :] = flat[-1:, :] = flat[:, :1] = flat[:, -1:] = False
    return flat


def match_flats(hi: np.ndarray, src: np.ndarray, tol: float = 3.0 / 255.0
                ) -> np.ndarray:
    """Put the upscaler's flat regions back on the source's colours.

    One affine map per channel, fitted from the upscaled image to a bicubic
    upsample of the same input, over the pixels the bicubic image says are flat.
    Uses nothing but the input, so it is available at inference; the artist's
    file is never consulted. Both arrays are in [0, 1].
    """
    from PIL import Image

    bi = np.asarray(
        Image.fromarray((src.clip(0, 1) * 255).astype(np.uint8)).resize(
            (hi.shape[1], hi.shape[0]), Image.BICUBIC), np.float64) / 255.0
    flat = _flat_mask(bi, tol)
    if flat.mean() < 0.01:  # nothing flat enough to fit on; use everything
        flat = np.ones(bi.shape[:2], bool)

    out = hi.astype(np.float64).copy()
    for c in range(3):
        x, y = hi[..., c][flat], bi[..., c][flat]
        if x.size < 16 or x.std() < 1e-6:
            continue
        A = np.stack([x, np.ones_like(x)], axis=1)
        a, b = np.linalg.lstsq(A, y, rcond=None)[0]
        out[..., c] = hi[..., c] * a + b
    return out.clip(0, 1)


def prepass(up, rgba: np.ndarray, out_scale: int = 2, recolour: bool = True,
            **kw) -> np.ndarray:
    """RGBA in [0, 1] -> cleaned RGBA at `out_scale` times the input size."""
    from . import model

    hi = model.upscale_rgba(up, rgba, **kw)
    factor = up.scale // out_scale
    if factor > 1:
        hi = box_downsample(hi, factor)
    elif factor < 1:
        raise ValueError(f"cannot output x{out_scale} from an x{up.scale} model")
    if recolour:
        # Fit against the RGB the model actually saw, not the raw file. Where a
        # PNG is transparent its colour channels hold whatever the encoder left
        # there, and `upscale_rgba` zeroes those before the network sees them;
        # matching against the un-zeroed original would fit the affine map to
        # pixels that never entered the upscale.
        src = np.where(rgba[..., 3:4] > 1e-4, rgba[..., :3], 0.0)
        hi = np.dstack([match_flats(hi[..., :3], src), hi[..., 3:4]])
    return hi.clip(0, 1)


# --------------------------------------------------------------------- detection
def interior_residual(inp_rgb: np.ndarray, model_rgb: np.ndarray) -> float:
    """RMS disagreement with the input, in 8-bit levels, where the trace is flat.

    Both arrays are the same size, in [0, 1], composited on white. The traced
    model is piecewise flat by construction, so wherever it is locally constant
    the input should be too. On an undamaged raster this is 0.000; JPEG q50 puts
    it at 1.509, which is the widest margin of any reference-free signal tried.

    Note that the cheaper candidate -- an 8x8 block signature, which needs no
    trace at all -- was measured and does not work: 2.175 on clean against 2.133
    on JPEG q50 is no signal. This one costs a first trace and is the one that
    discriminates.
    """
    flat = _flat_mask(model_rgb, 1.5 / 255.0)
    if flat.sum() < 100:
        return float("nan")
    d = (model_rgb - inp_rgb)[flat] * 255.0
    return float(np.sqrt((d ** 2).mean() / 3))

File: crates/inkvec-sr/src/clean.rs
//! The pre-pass arithmetic: box downsample, and putting the colours back.
//!
//! The upscaler removes about two thirds of a JPEG's error energy -- not because it was trained to, it was
//! not, but because it was fitted on flat-colour logo art and ringing is not on
//! that manifold, so it cannot represent the damage.
//!
//! Three steps, each here for a measured reason.
//!
//! **Upscale x4** is what cleans. Artefact gain -- what survives the network
//! relative to what went in -- is 0.36 across seven families.
//!
//! **Box-downsample x2** is not a denoiser, whatever it looks like. A 2x2 box
//! has its only transfer zero at Nyquist, and the DCT grid puts JPEG damage at
//! f = 1/8 cycles/pixel and below, which it passes at 0.974 of amplitude: half
//! the error survives, and the surviving half is the *more* coherent half. It is
//! here because x2 is the useful output scale, not because it cleans.
//!
//! **Recolour** exists because the network costs ~0.6 dE00 on clean input too.
//! Split by region against bicubic as a control, that is 0.60 on flat interiors
//! where bicubic scores 0.03, and 1.57 on edges where bicubic scores 8.63. Each
//! resampler is right about the half the other is wrong about, so take geometry
//! from the network and colour from the source.

use inkvec_trace::Rgba;

/// Exact box average by an integer factor. Alpha rides along with the colour.
pub fn box_downsample(img: &Rgba, factor: usize) -> Rgba {
    assert!(factor >= 1, "factor must be positive");
    if factor == 1 {
        return img.clone();
    }
    let (w, h) = (img.width / factor, img.height / factor);
    let mut data = vec![0.0f32; w * h * 4];
    let inv = 1.0 / (factor * factor) as f32;
    for y in 0..h {
        for x in 0..w {
            let mut acc = [0.0f32; 4];
            for dy in 0..factor {
                for dx in 0..factor {
                    let p = img.pixel(x * factor + dx, y * factor + dy);
                    for c in 0..4 {
                        acc[c] += p[c];
                    }
                }
            }
            let o = (y * w + x) * 4;
            for c in 0..4 {
                data[o + c] = acc[c] * inv;
            }
        }
    }
    Rgba {
        width: w,
        height: h,
        data,
    }
}

/// Bicubic resample to an explicit size. Catmull-Rom (a = -0.5), matching what
/// PIL calls BICUBIC, because the recolour map is fitted against it and a
/// different kernel would move the flat mask.
pub fn bicubic(img: &Rgba, w: usize, h: usize) -> Rgba {
    fn weight(t: f32) -> f32 {
        const A: f32 = -0.5;
        let t = t.abs();
        if t <= 1.0 {
            (A + 2.0) * t * t * t - (A + 3.0) * t * t + 1.0
        } else if t < 2.0 {
            A * (t * t * t - 5.0 * t * t + 8.0 * t - 4.0)
        } else {
            0.0
        }
    }
    let mut data = vec![0.0f32; w * h * 4];
    let sx = img.width as f32 / w as f32;
    let sy = img.height as f32 / h as f32;
    for y in 0..h {
        let fy = (y as f32 + 0.5) * sy - 0.5;
        let iy = fy.floor() as isize;
        for x in 0..w {
            let fx = (x as f32 + 0.5) * sx - 0.5;
            let ix = fx.floor() as isize;
            let mut acc = [0.0f32; 4];
            let mut wsum = 0.0f32;
            for m in -1..=2isize {
                let wy = weight(fy - (iy + m) as f32);
                if wy == 0.0 {
                    continue;
                }
                let yy = (iy + m).clamp(0, img.height as isize - 1) as usize;
                for n in -1..=2isize {
                    let wx = weight(fx - (ix + n) as f32);
                    if wx == 0.0 {
                        continue;
                    }
                    let xx = (ix + n).clamp(0, img.width as isize - 1) as usize;
                    let p = img.pixel(xx, yy);
                    let ww = wx * wy;
                    for c in 0..4 {
                        acc[c] += p[c] * ww;
                    }
                    wsum += ww;
                }
            }
            let o = (y * w + x) * 4;
            let inv = if wsum.abs() > 1e-8 { 1.0 / wsum } else { 0.0 };
            for c in 0..4 {
                data[o + c] = (acc[c] * inv).clamp(0.0, 1.0);
            }
        }
    }
    Rgba {
        width: w,
        height: h,
        data,
    }
}

/// True where the 3x3 neighbourhood of the luma spans less than `tol`. Border
/// pixels are never flat: they have no full neighbourhood to be judged on.
pub fn flat_mask(img: &Rgba, tol: f32) -> Vec<bool> {
    let (w, h) = (img.width, img.height);
    let luma: Vec<f32> = (0..w * h)
        .map(|i| {
            let p = &img.data[i * 4..i * 4 + 3];
            (p[0] + p[1] + p[2]) / 3.0
        })
        .collect();
    let mut mask = vec![false; w * h];
    if w < 3 || h < 3 {
        return mask;
    }
    for y in 1..h - 1 {
        for x in 1..w - 1 {
            let (mut lo, mut hi) = (f32::INFINITY, f32::NEG_INFINITY);
            for dy in 0..3 {
                for dx in 0..3 {
                    let v = luma[(y + dy - 1) * w + (x + dx - 1)];
                    lo = lo.min(v);
                    hi = hi.max(v);
                }
            }
            mask[y * w + x] = hi - lo < tol;
        }
    }
    mask
}

/// Put the upscaler's flat regions back on the source's colours.
///
/// One affine map per channel, fitted from the upscaled image to a bicubic
/// upsample of the same input, over the pixels the bicubic image says are flat.
/// Uses nothing but the input, so it is available at inference; the artist's
/// file is never consulted.
pub fn match_flats(hi: &mut Rgba, src: &Rgba) {
    const TOL: f32 = 3.0 / 255.0;
    let bi = bicubic(src, hi.width, hi.height);
    let mut mask = flat_mask(&bi, TOL);
    let n_flat = mask.iter().filter(|&&m| m).count();
    if (n_flat as f64) < 0.01 * mask.len() as f64 {
        // Nothing flat enough to fit on; use every pixel rather than nothing.
        mask.iter_mut().for_each(|m| *m = true);
    }

    for c in 0..3 {
        // Least squares for y = a*x + b over the masked pixels, in f64: the
        // normal equations of a two-parameter fit over ~10^5 samples lose more
        // than f32 has to give.
        let (mut sx, mut sy, mut sxx, mut sxy, mut n) = (0.0f64, 0.0f64, 0.0f64, 0.0f64, 0u64);
        for (i, &m) in mask.iter().enumerate() {
            if !m {
                continue;
            }
            let x = hi.data[i * 4 + c] as f64;
            let y = bi.data[i * 4 + c] as f64;
            sx += x;
            sy += y;
            sxx += x * x;
            sxy += x * y;
            n += 1;
        }
        if n < 16 {
            continue;
        }
        let nf = n as f64;
        let det = nf * sxx - sx * sx;
        if det.abs() < 1e-9 {
            continue; // a constant channel carries no slope to fit
        }
        let a = (nf * sxy - sx * sy) / det;
        let b = (sy - a * sx) / nf;
        for i in 0..hi.width * hi.height {
            let v = hi.data[i * 4 + c] as f64;
            hi.data[i * 4 + c] = ((a * v + b) as f32).clamp(0.0, 1.0);
        }
    }
}

/// Composite onto white, dropping alpha.
pub fn on_white(img: &Rgba) -> Vec<[f32; 3]> {
    (0..img.width * img.height)
        .map(|i| {
            let p = &img.data[i * 4..i * 4 + 4];
            [
                p[0] * p[3] + (1.0 - p[3]),
                p[1] * p[3] + (1.0 - p[3]),
                p[2] * p[3] + (1.0 - p[3]),
            ]
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn solid(w: usize, h: usize, c: [f32; 4]) -> Rgba {
        Rgba {
            width: w,
            height: h,
            data: (0..w * h).flat_map(|_| c).collect(),
        }
    }

    #[test]
    fn box_averages_exactly() {
        let mut img = solid(4, 4, [0.0, 0.0, 0.0, 1.0]);
        // Red channel of the pixel at (x, y) on a 4-wide RGBA image.
        let red = |x: usize, y: usize| (y * 4 + x) * 4;
        // One 2x2 block set to 1.0 in red, so its average must be exactly 1.0,
        // and a block with a single 1.0 must average to exactly 0.25.
        for (x, y) in [(0, 0), (1, 0), (0, 1), (1, 1)] {
            img.data[red(x, y)] = 1.0;
        }
        img.data[red(2, 0)] = 1.0;
        let d = box_downsample(&img, 2);
        assert_eq!(d.width, 2);
        assert!((d.pixel(0, 0)[0] - 1.0).abs() < 1e-6);
        assert!((d.pixel(1, 0)[0] - 0.25).abs() < 1e-6);
    }

    #[test]
    fn bicubic_preserves_a_constant() {
        let img = solid(8, 8, [0.25, 0.5, 0.75, 1.0]);
        let up = bicubic(&img, 32, 32);
        for i in 0..up.width * up.height {
            assert!((up.data[i * 4] - 0.25).abs() < 1e-5, "at {i}");
            assert!((up.data[i * 4 + 2] - 0.75).abs() < 1e-5, "at {i}");
        }
    }

    #[test]
    fn recolour_inverts_a_known_bias() {
        // A flat source, and an "upscale" of it carrying a linear distortion.
        // The fit must undo the distortion exactly.
        let mut src = solid(16, 16, [0.4, 0.4, 0.4, 1.0]);
        for i in 0..16 * 16 {
            let v = 0.2 + 0.6 * ((i % 16) as f32 / 15.0);
            src.data[i * 4] = v;
            src.data[i * 4 + 1] = v;
            src.data[i * 4 + 2] = v;
        }
        let mut hi = bicubic(&src, 32, 32);
        for i in 0..32 * 32 {
            for c in 0..3 {
                hi.data[i * 4 + c] = (hi.data[i * 4 + c] * 0.9 + 0.05).clamp(0.0, 1.0);
            }
        }
        let want = bicubic(&src, 32, 32);
        match_flats(&mut hi, &src);
        let err = (0..32 * 32)
            .map(|i| (hi.data[i * 4] - want.data[i * 4]).abs())
            .fold(0.0f32, f32::max);
        assert!(err < 2e-2, "residual bias after recolour: {err}");
    }

    #[test]
    fn flat_mask_finds_the_flat_part() {
        let mut img = solid(16, 16, [0.5, 0.5, 0.5, 1.0]);
        for y in 0..16 {
            for x in 8..16 {
                for c in 0..3 {
                    img.data[(y * 16 + x) * 4 + c] = if x % 2 == 0 { 0.0 } else { 1.0 };
                }
            }
        }
        let m = flat_mask(&img, 3.0 / 255.0);
        assert!(m[8 * 16 + 3], "the constant half should be flat");
        assert!(!m[8 * 16 + 11], "the striped half should not be");
    }
}
